### PR TITLE
[DX] Add ValidateFixtureNamespaceCommand

### LINKED ISSUE #4065: [DX] Add custom fixture check for namespace
- make sure the fixture file is in correct namespace
- the class inside name respects the file name, e.g `Fixture/some_file.php.inc` must have `class SomeFile` in it

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes #4065

### CODE REVIEW

**Comment by samsonasik** [comment]
Currently checked against `/tests` root directory, got 15 files with invalid namespace:

```bash
bin/rector validate-fixture-namespace 
Rector dev-master@61828cd
Config file: rector.php


 * /Users/samsonasik/www/rector/tests/Exclusion/Check/Fixture/docblock-on-parent-baseline.php.inc
 * /Users/samsonasik/www/rector/tests/Exclusion/Check/Fixture/docblock-on-parent-norector.php.inc
 * /Users/samsonasik/www/rector/tests/Exclusion/Check/Fixture/docblock-on-self-baseline.php.inc
 * /Users/samsonasik/www/rector/tests/Exclusion/Check/Fixture/docblock-on-self-no-rector.php.inc
 * /Users/samsonasik/www/rector/tests/Exclusion/Check/Fixture/other-docblocks.php.inc
 * /Users/samsonasik/www/rector/tests/Issues/Issue4499/DoNotDuplicateCode/Fixture/has_interface_exists_check.php.inc
 * /Users/samsonasik/www/rector/tests/Issues/Issue4499/DoNotDuplicateCode/Fixture/has_interface_exists_check_without_bracket.php.inc
 * /Users/samsonasik/www/rector/tests/Issues/Issue4499/DoNotDuplicateCode/Fixture/has_trait_exists_check.php.inc
 * /Users/samsonasik/www/rector/tests/Issues/Issue4499/DoNotDuplicateCode/Fixture/has_trait_exists_check.php_without_bracket.php.inc
 * /Users/samsonasik/www/rector/tests/Issues/Issue4543/DoNotFailOnEmptyFile/Fixture/empty_file.php.inc
 * /Users/samsonasik/www/rector/tests/Issues/Issue4581/DoNotChangeAnnotationColumn/Fixture/skip_annotation_column.php.inc
 * /Users/samsonasik/www/rector/tests/PhpParser/Printer/CommentRemover/Fixture/asterisk_comment.php.inc
 * /Users/samsonasik/www/rector/tests/PhpParser/Printer/CommentRemover/Fixture/behind_value_comment.php.inc
 * /Users/samsonasik/www/rector/tests/PhpParser/Printer/CommentRemover/Fixture/cross_bracket_comment.php.inc
 * /Users/samsonasik/www/rector/tests/PhpParser/Printer/CommentRemover/Fixture/many_comment.php.inc

                                                                                                                        
 [ERROR] Found 15 files with invalid namespace. Please follow psr-4 defined in composer.json  
```

I will patch the namespace of files under `tests` first before continue check `packages` and `rules`.

**Comment by samsonasik** [comment]
fixed fixture namespace in tests directory

**Comment by samsonasik** [comment]
@TomasVotruba it `found 689 files with invalid namespace` in `rules` directory.

**Comment by samsonasik** [comment]
Screen Shot 2020-12-10 at 11 45 24 PM

**Comment by samsonasik** [comment]
@TomasVotruba I think it need `--fix` option to allow fix it in the files as there are hundreds of invalid namespace in fixture files.

**Comment by samsonasik** [comment]
@TomasVotruba tests fixture fixed. Failure tests case seems unrelated:

```bash
626) Rector\DeadCode\Tests\Rector\ClassMethod\RemoveUnusedPrivateMethodRector\RemoveUnusedPrivateMethodRectorTest::test with data set #5 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
_HumbugBoxfb21822734fc\Nette\DI\ServiceCreationException: Service '0286': Class Symplify\PHPStanRules\Rules\NoParticularNodeRule not found.
```

May wait for https://github.com/rectorphp/rector/pull/4848

**Comment by samsonasik** [comment]
rebased.

**Comment by samsonasik** [comment]
Current `--fix` functionality only for namespace. Many class name conflicts after namespace updated, need to update class name as well with `--fix` to avoid manual change next.

**Comment by samsonasik** [comment]
I am adding new command `ValidateFixtureClassNameCommand` to fix the class name.

**Comment by samsonasik** [comment]
so many false positive, need to exclude more paths

**Comment by samsonasik** [comment]
It really needs to one by one rules directory check for classname fixture fix.

File: tests/Exclusion/Check/Fixture/docblock-on-parent-norector.php.inc [hunk 1]
```diff
@@ -1,9 +1,9 @@
 <?php
 
-namespace Rector\Core\Tests\Rector\Exclusion\Check\DocBlockOnParentNoRector;
+namespace Rector\Core\Tests\Exclusion\Check\Fixture;
 
 
-final class SomeController
+final class Docblock-on-parent-norector
```
Comment (samsonasik) [inline tests/Exclusion/Check/Fixture/docblock-on-parent-norector.php.inc:6]: Oh, dash separated fixture file

Comment (samsonasik) [reply tests/Exclusion/Check/Fixture/docblock-on-parent-norector.php.inc]: fixed.

**Comment by samsonasik** [comment]
The `validate-fixture-namespace` and `validate-fixture-classname` --fix fixture applied for `architecture` rule, will apply next rules one by one.

**Comment by simivar** [comment]
What do you think about splitting this PR into a few separate ones so it's easier to review?

For example:
1. adding commands
2. fixing problems in the current codebase
3. adding commands to GitHub Actions

**Comment by samsonasik** [comment]
The false positive mostly in rules part, I think we can apply to tests and/or packages directories first

**Comment by TomasVotruba** [comment]
@samsonasik I agree with @simivar . 
Having bare commands first would easy the further edge-case fixes process.

Could you decouple first few commits withoutt testing in CI to own pull-request? We could merge that already and narrow the scope.

**Comment by samsonasik** [comment]
I will try.

**Comment by TomasVotruba** [comment]
@samsonasik How should we proceed with this PR?

**Comment by samsonasik** [comment]
It need a one by one add rules directory to directory to be checked in validate-fixture-namespace/classname. I will try create new PR for it.

**Comment by TomasVotruba** [comment]
Ok, thank you. This PR can be closed then.
We'll process the small ones, one by one
